File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/54-SpiralMatrix/54-SpiralMatrix.java
// Last updated: 8/4/2026, 7:24:21 PM
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer>ll = new ArrayList<>();
        int n = matrix.length;
        int m = matrix[0].length;

        int top = 0 , left = 0 , bottom = n-1 , right = m-1;

        while(top <= bottom && left <= right){

            for(int i = left ; i <= right ; i++){
                ll.add(matrix[top][i]);
            }
            top++;

            for(int i = top ; i <= bottom ; i++){
                ll.add(matrix[i][right]);
            }
            right--;

            if(top <= bottom){
                for(int i = right ; i >= left ; i--){
                    ll.add(matrix[bottom][i]);
                }
                bottom--;
            }
            

            if(left <= right){
                for(int i = bottom ; i >= top ; i--){
                    ll.add(matrix[i][left]);
                }
                left++;
            }
        }
        return ll;
    }
}

File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/54-SpiralMatrix/54-SpiralMatrix.cpp
// Last updated: 8/4/2026, 7:24:27 PM
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<int>result;
        int top = 0, bottom = m - 1 , left = 0 , right = n - 1;

        while(left <= right && top <= bottom){

            for(int i = left ; i <=right ; i++){
                result.push_back(matrix[top][i]);
            }
            top++;
            
            for(int i = top ; i <=bottom ; i++){
                result.push_back(matrix[i][right]);
            }
            right--;

            if(top <= bottom){
            for(int i = right ; i >= left ; i--){
                result.push_back(matrix[bottom][i]);
            }
            bottom--;
            }

            if(left <= right){
            for(int i = bottom ; i >=top ; i--){
                result.push_back(matrix[i][left]);
            }
            left++;
            }
        }
        return result;
    }
};